storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -93,7 +92,8 @@
   final CachingTaskQueue<Bytes32, StateAndBlockSummary> states;
   final Map<Bytes32, SignedBeaconBlock> blocks;
   final CachingTaskQueue<SlotAndBlockRoot, BeaconState> checkpointStates;
-  final Map<UInt64, VoteTracker> votes;
+  final VoteTracker[] votes;
+  public static final int VOTE_TRACKER_SIZE = 1024;
```

Comment:
This actually can't be a fixed size array - there will be one VoteTracker for each validator so the array will have to grow as the validator set grows.

The best way to achieve that is probably to actually use an `ArrayList` - it still stores the entries in an array but manages expanding the array as necessary for you.

So the declaration would be:

```
final List<VoteTracker> votes = new ArrayList<>();
```

access would be via the `set(index, value)` and `get(index)` methods.

Reply:
@ajsutton  
q1) 
In HashMap it doesn't guarantee (key, value ) traversal in order of key inserts(LinkedHashMap has this guarantee) ,  if we use `set(index, value) ` directly it will cause null pointer exception if there is no existing node there, so we have to pre-populate the array to the right size with nulls `votes.add(null) `. To prepropulate the array we need to know the size, so we need to compute the HighestVotedValidator index apriori and then populate that many entries in this array.  so clubbing max validator index along with populating ArrayList won't work.   

q2)
`dumb doubt`
In the `public static UpdatableStore create()` method we pass `final Map<UInt64, VoteTracker> votes` , can I assume that if I sort  keys of this HashMap then they are consecutive, and lie in some ` [0.. SIZE] ` range , and a key exist for each index (no missing  keys in this range ).  I don't think my statement is true as I see nulls are expected during getVote() 

```java
  @Override
  public VoteTracker getVote(UInt64 validatorIndex) {
    VoteTracker txVote = votes.get(validatorIndex.intValue());
    if (txVote != null) {
      return txVote;
    } else {
      VoteTracker storeVote = store.getVote(validatorIndex);
      return storeVote != null ? storeVote : VoteTracker.DEFAULT;
    }
  }
```
 
q3) 
can you confirm if this is correct. 
```java
//existing commit() code in StoreVoteUpdator.java 

 @Override
  public void commit() {
    // Votes are applied to the store immediately since the changes to the in-memory ProtoArray
    // can't be rolled back.
    store.votes.putAll(votes);
    store.highestVotedValidatorIndex =
        store.highestVotedValidatorIndex.max(
            votes.keySet().stream().max(Comparator.naturalOrder()).orElse(UInt64.ZERO));
    voteUpdateChannel.onVotesUpdated(votes);
  }
  ```
  
  ```java
 // after rewriting it assuming votes are of type ArrayList, code becomes like this. 
   public void commit() {
    // Votes are applied to the store immediately since the changes to the in-memory ProtoArray
    // can't be rolled back.
    
    store.highestVotedValidatorIndex =
        store.highestVotedValidatorIndex.max(
            votes.keySet().stream().max(Comparator.naturalOrder()).orElse(UInt64.ZERO));
    while(store.votes.size() <=store.highestVotedValidatorIndex)store.votes.add(null);
    
    
    //store.votes.putAll(votes); gets translated to this 
    votes.forEach(
        (key, value) -> {
          store.votes.set(key.intValue(),value) ;
        });
            
    HashMap<UInt64, VoteTracker> hs = new HashMap<>();
    for(int i=0;i<(int)store.votes.size();++i){
        if(store.votes[i]!=null){
            hs.put(UInt64.valueOf(i), store.votes.get(i));
        }
    }
    
    voteUpdateChannel.onVotesUpdated(hs); //this is so that we keep api(contract) unchanged for VoteUpdateChannel
  }

 ```

Reply:
> @ajsutton
> q1)
> In HashMap it doesn't guarantee (key, value ) traversal in order of key inserts(LinkedHashMap has this guarantee) , if we use `set(index, value) ` directly it will cause null pointer exception if there is no existing node there, so we have to pre-populate the array to the right size with nulls `votes.add(null) `. To prepropulate the array we need to know the size, so we need to compute the HighestVotedValidator index apriori and then populate that many entries in this array. so clubbing max validator index along with populating ArrayList won't work. unless we assume HashMap.size() == highestVotedKeyValidatorIndex.

Ah yes, I had forgotten that `ArrayList.set` didn't just automatically grow the list.  In that case it probably winds up being to just use a plain `VoteTracker[]` again and manually create a bigger one when necessary.  The `Store.highestVotedValidatorIndex` variable tracks the highest validator number that has voted and thus the minimum size of array we'd need. Note though that as more validators activate that number may increase and we'll need to increase the size of the array to fit them all (eg with `Arrays.copyOf`).  To avoid copying it too often, when we do copy it we should give ourselves a bit of extra space.  About 900 validators can be activated a day so I'd suggest we leave 1000 empty slots when we need to ensure it.  That's a maximum wastage of about 0.5% but means we only perform the relatively expensive copy about once a day.

> q2)
> `dumb doubt`
> In the `public static UpdatableStore create()` method we pass `final Map<UInt64, VoteTracker> votes` , can I assume that if I sort keys of this HashMap then they are consecutive, and lie in some `[0.. SIZE]` range , and a key exist for each index (no missing keys in this range ). I don't think my statement is true as I see nulls are expected during getVote()

No you can' assume that.  In fact it's pretty much guaranteed to be a random assortment of indices in there.  But I'm not sure why you're needing to sort them.  It isn't replacing all the votes in the store, just doing a selective update.  So any vote in that map overwrites the previous vote from that validator in `Store.votes` and the other entries in `Store.votes` are left unchanged.

> q3)
> can you confirm if this is correct. code changes for commit() in StoreVoteUpdator.java class
> 
> ```java
> //existing commit() code in StoreVoteUpdator.java 
> 
>  @Override
>   public void commit() {
>     // Votes are applied to the store immediately since the changes to the in-memory ProtoArray
>     // can't be rolled back.
>     store.votes.putAll(votes);
>     store.highestVotedValidatorIndex =
>         store.highestVotedValidatorIndex.max(
>             votes.keySet().stream().max(Comparator.naturalOrder()).orElse(UInt64.ZERO));
>     voteUpdateChannel.onVotesUpdated(votes);
>   }
> ```
> 
>  ```java
> // after rewriting it assuming votes are of type ArrayList, code becomes like this. 
>  public void commit() {
>   // Votes are applied to the store immediately since the changes to the in-memory ProtoArray
>   // can't be rolled back.
>   
>   store.highestVotedValidatorIndex =
>       store.highestVotedValidatorIndex.max(
>           votes.keySet().stream().max(Comparator.naturalOrder()).orElse(UInt64.ZERO));
>   while(store.votes.size() <=store.highestVotedValidatorIndex)store.votes.add(null);
>   
>   
>   //store.votes.putAll(votes); gets translated to this 
>   votes.forEach(
>       (key, value) -> {
>         store.votes.set(key.intValue(),value) ;
>       });
> ```

Up to here is correct, though `while(store.votes.size() <=store.highestVotedValidatorIndex)store.votes.add(null);` is too inefficient to do in practice.  This is where we'd grow the size of the array as discussed above.

> ```
>   HashMap<UInt64, VoteTracker> hs = new HashMap<>();
>   for(int i=0;i<(int)store.votes.size();++i){
>       if(store.votes[i]!=null){
>           hs.put(UInt64.valueOf(i), store.votes.get(i));
>       }
>   }
>   
>   voteUpdateChannel.onVotesUpdated(hs); //this is so that we keep api(contract) unchanged for VoteUpdateChannel
> }
> ```

This merging of store votes isn't required.  Only the changed votes need to be set to storage via `voteUpdateChannel` so the existing `voteUpdteChannel.onVotesUpdated(votes)` is still all that's required.

Reply:
@ajsutton I have a suggestion , let's  remove the below two fields from StoreVoteUpdater 
```java
  private final Map<UInt64, VoteTracker> votes = new HashMap<>();
  private UInt64 highestVotedValidatorIndex = UInt64.ZERO;
```

it makes the separation of concerns(Single Responsibility Principle) clear . 

for example, this will make the below code 

```java
  @Override
  public VoteTracker getVote(UInt64 validatorIndex) {
    VoteTracker txVote = votes.get(validatorIndex);
    if (txVote != null) {
      return txVote;
    } else {
      VoteTracker storeVote = store.getVote(validatorIndex);
      return storeVote != null ? storeVote : VoteTracker.DEFAULT;
    }
  }
```

to this 

```java
  @Override
  public VoteTracker getVote(UInt64 validatorIndex) {

      VoteTracker storeVote = store.getVote(validatorIndex);
      return storeVote != null ? storeVote : VoteTracker.DEFAULT;
    
  }
```

now it will be clear and updating the  highestVotedValidatorIndex  will be the responsibility of Store class alone.  this will help in removing code duplication as well.

Reply:
> @ajsutton
> 
> 1. can we   remove the below two fields from `StoreVoteUpdater` ?
> 
> ```java
>   private final Map<UInt64, VoteTracker> votes = new HashMap<>();
>   private UInt64 highestVotedValidatorIndex = UInt64.ZERO;
> ```
> 
> it keeps the `separation of concerns` clear .

The `StoreVoteUpdater` class is providing transaction isolation. So if in the process of updating all the votes we hit an exception, the `StoreVoteUpdater` never gets committed and all changes are dropped. That way we wind up with either apply all changes or none.  That's why `StoreVoteUpdater` maintains it's own `votes` and `highestVotedValidatorIndex` so that they are consistent within that transaction, but not applied to the `Store` until `commit()` is called.

So we'd need to keep them separate.  If you think about it as `Store` being an in-memory version of a transactional database it helps to get the mental model straight.


> 1. is there any reason for  having locks on `Store.getVote()` but not on `StoreVoteUpdater.getVote()` ? if explanation is bit long you can ignore this question.
> 
> my guess: `Store` object is a singleton, where as `StoreVoteUpdater` objects are created once per every new thread, why Store object has to be singleton, that i don't know.


Essentially yes - there is one instance of `Store` shared by the application (it's not strictly a singleton in terms of the design pattern, you can create two instances of it but it is a shared object).  Having the `Store` shared is important so that each part of the system has a consistent view of the state of the chain.  Again, think in-memory database - operations on non-finalised blocks don't generally need to access the disk, they just operate on the data held in the `Store` instance.  If memory is tight, `Store` can drop things out of memory and then reload them from disk but that's transparent to the rest of the application.  As far as Teku is concerned all the data for the non-finalised part of the chain is available from the `Store` instance.

And then yes `StoreVoteUpdater` is created for each transaction and so doesn't need to be thread-safe and can skip the locks.


>  1. why are we not having a lock while doing `store.votes.putAll(votes)` in the below method at `StoreVoteUpdater `class ?

hmm, that's an excellent question. I suspect that's a bug - we're likely getting away with it because there is actually only a single thread that ever accesses or updates votes but it definitely doesn't fit with the threading model.  I'll need to do some more analysis to make sure there isn't a potential for deadlock by introducing the lock here but I think it probably should be there.


>  1. If we proceed with having `votes` instance field as of type `VoteTracker[]` in `StoreVoteUpdater` class.
>    the `commit()` method transitions from this, is the modified method correct ?
> ```java
>   @Override
> public void commit() {
>   // Votes are applied to the store immediately since the changes to the in-memory ProtoArray
>   // can't be rolled back.
>   
>   store.highestVotedValidatorIndex =
>       store.highestVotedValidatorIndex.max(
>           votes.keySet().stream().max(Comparator.naturalOrder()).orElse(UInt64.ZERO));

It occurs to me that we already have `StoreVoteUpdater.highestVotedValidatorIndex` tracking the highest modified vote so we could simplify this line to:

```
store.highestVotedValidatorIndex = getHighestVotedValidatorIndex();
```



>     if ( store.highestVotedValidatorIndex.intValue() >= store.votes.length )  {
>          store.votes = Arrays.copyOf(store.votes, store.highestVotedValidatorIndex.intValue()+1000);
>     } 
>     
>     
>     //store.votes.putAll(votes);
>     for(int i=0;i<votes.length;++i)store.votes[i]=votes[i];
>    
>    HashMap<UInt64, VoteTracker>updatedVotes = new HashMap<>();
>    for(int i=0;i<votes.length;++i){
>        if(votes[i]!=null){
>           updatedVotes.put(UInt64.valueOf(i), store.votes.get(i));
>        }
>    }
> 
>   voteUpdateChannel.onVotesUpdated(updatedVotes);


I think we'd want to keep `StoreVoteUpdater.votes` as a `HashMap` since most votes won't be updated in a single transaction. It's only `Store.votes` that we want to change. So this would just be:

```
updatedVotes.forEach((index, vote) -> store.votes[index.intValue()] = vote);
voteUpdateChannel.onVotesUpdated(votes);
```

Reply:
> store.highestVotedValidatorIndex = getHighestVotedValidatorIndex();

done

Reply:
> I think we'd want to keep StoreVoteUpdater.votes as a HashMap since most votes won't be updated in a single transaction. It's only Store.votes that we want to change. So this would just be:

got it , will not update the type of  StoreVoteUpdater.votes